scripts/calculate_trending_scores_hourly.py

```diff
@@ -0,0 +1,124 @@
+from openlibrary.plugins.worksearch.code import execute_solr_query
+from math import sqrt
+
+# This script handles hourly updating of each works' z-score. The 'trending_score_hourly_23' field is
+# ignored, and the current count of bookshelves events in the last hour is the new trending_score_hourly[0]
+# value, with each other value incrementing up by one.
+
+# Trending_score_daily values are all fetched in order to calculate their mean and standard deviation.
+# The formula for the new z-score is [ z = sum(trending_score_hourly_*) - mean(trending_score_daily_*) /  (standard_deviation(trending_score_daily)] + 1).
+# The incrementation of the denominator by 1 serves several purposes:
+#       - It prevents divide-by-zero errors on objects with 0s in all trending_score_daily_* values, which is important,
+#         as they are instantiated to that upon the first time solr reindexes them with this new information.
+#       - It serves to deemphasize such values, as we're not particularly interested in a shift from
+#         0 events to 1 event.
+
+
+def fetch_works(current_hour: int):
+    ol_db = db.get_db()
+    query = "select work_id, Count(updated)"
+    query += "from bookshelves_events group by bookshelves_events.updated, work_id "
+    query += "having updated >= localtimestamp - interval '1 hour'"
+    db_data = {
+        f'/works/OL{storage.work_id}W': storage.count
+        for storage in list(ol_db.query(query))
+    }
+    print(db_data)
+    query = f'trending_score_hourly_{current_hour}:[1 TO *] OR {" OR ".join(["key:\"" + key + "\"" for key in db_data])}'
+    print(query)
+    resp = execute_solr_query(
+        '/export',
+        {
+            "q": query,
+            "fl": ",".join(
+                [
+                    "key",
+                    "trending_score_hourly_sum",
+                    f'trending_score_hourly_{current_hour}',
+                    "trending_score_daily_*",
+                ]
+            ),
+            "sort": "trending_score_hourly_sum desc",
+        },
+    )
+    doc_data = {}
+    if resp:
+        data = resp.json()
+        try:
+            docs = data["response"]["docs"]
+
+        except KeyError:
+            raise KeyError
+        print(docs)
+        doc_data = {
+            doc["key"]: {"count": db_data.get(doc["key"], 0), "solr_doc": doc}
+            for doc in docs
+        }
+    return doc_data
+
+
+def get_z_score(solr_doc: dict, count: int, current_hour: int):
+    arith_mean = sum([solr_doc[f'trending_score_daily_{i}'] for i in range(7)]) / 7
+    last_24_hours_value = (
+        solr_doc['trending_score_hourly_sum']
+        + count
+        - solr_doc[f'trending_score_hourly_{current_hour}']
+    )
+    st_dev = sqrt(
+        sum(
+            [
+                pow(solr_doc[f'trending_score_daily_{i}'] - arith_mean, 2)
+                for i in range(7)
+            ]
+        )
+        / 7
+    )
+    return (last_24_hours_value - arith_mean) / (st_dev + 1.0)
```

Comment:
Where is this value thresholded, to indicate detection of a change?

Reply:
I'm not entirely sure what you mean by this.  Could you clarify?

Reply:
I see, so instead of setting a cutoff among z-scores, you set a cutoff in the ranking of z-scores.  What I'll do is build a simple simulation to estimate the errors - the spurious and overlooked results due to randomness, so the method can be tuned

Reply:
This graph should give some insight.  It shows that the percentage to be "marked" does not depend on the event rate - so rarely touched items would appear with the same frequency as popular ones.  Perhaps this is what one wants?  (Ignore the jaggyness)
![image](https://github.com/user-attachments/assets/1af9ac91-3e37-4750-aa51-0c7b3d6c7720)

Reply:
To a degree, this is what we want from the data. I've been running some tests on my own end, though I didn't make it to any particular graphs. 
I've proposed a few cutoffs on my own, to only assign a z-score higher than zero if there have been at least 'x' number  of events in the past week,  as at the moment, our measurement seem to be  drawing more of the 1/10 results as desired-- specifically, there are a lot of cases where no results for a week, and then a small amount ~(2<5) on the final day drives the work straight up to the top of the list.   

My current thoughts are that, in order to mark something as trending, it should have to average out to at least 1 event per day, over the last week; I don't feel that something with less than 5 entries over 7 days deserves  top of the list treatment, given the massive amount of works that have a very low event frequency. This is all on my own scratch code though, rn, rather than on the pull request; I haven't had the chance to discuss the proposal with Drini just yet.